File: crates/pdftract-py/tests/smoke_test.py
#!/usr/bin/env python3
"""Basic smoke test for pdftract SDK.

This test verifies that the pdftract module can be imported and basic
operations work correctly. It serves as a quick sanity check that the
SDK is properly structured and functional.

Usage:
    python3 smoke_test.py

The test uses minimal PDF fixtures to keep execution fast and reliable.
"""

from __future__ import annotations

import sys
from pathlib import Path

# Add the python package to the path
sys.path.insert(0, str(Path(__file__).parent.parent / "python"))

import pdftract
from pdftract import Document, Page, Span


def test_extract_returns_typed_document() -> None:
    """Verify extract() returns a typed Document instance.

    This smoke test validates the core type contract:
    - Document.from_native() returns a properly typed Document instance
    - Document has pages and metadata attributes with correct types
    - Page objects are properly typed with structural attributes (page, width, height)
    - Span objects are properly typed with expected attributes (text, bbox, font, size)
    - Nested structure integrity: Document -> Pages -> Spans hierarchy is complete
    - Content verification: spans contain real (non-empty) text content
    - Count integrity: all objects are properly accounted for

    Assertion types covered:
    - isinstance() checks for type verification
    - hasattr() checks for attribute existence
    - Length/content checks for data validation
    - Type-specific checks (str, tuple, numeric types for span attributes)

    This test loads fixture data from test-minimal.expected.json
    which contains actual page content for type verification.
    """
    # Load fixture data that has actual pages and spans
    fixture_path = Path(__file__).parent.parent.parent.parent / "tests" / "fixtures" / "test-minimal.expected.json"

    if not fixture_path.exists():
        print(f"❌ Fixture not found: {fixture_path}")
        sys.exit(1)

    # ===== FIXTURE LOADING =====
    # Load fixture data and create Document instance for testing
    import json
    try:
        with fixture_path.open("r") as f:
            fixture_data = json.load(f)
    except json.JSONDecodeError as e:
        print(f"❌ Fixture file contains invalid JSON: {e}")
        sys.exit(1)
    except Exception as e:
        print(f"❌ Failed to read fixture file: {e}")
        sys.exit(1)

    doc = pdftract.Document.from_native(fixture_data)

    # ===== Document-level type verification =====
    # Verify the top-level Document object is properly typed and has expected attributes

    assert isinstance(doc, pdftract.Document), \
        f'Expected Document instance from Document.from_native(), got {type(doc).__name__}'
    print("✓ Document.from_native() returns Document instance")

    assert hasattr(doc, 'pages'), \
        f"Document instance should have 'pages' attribute (type: {type(doc).__name__})"
    print("✓ Document has 'pages' attribute")

    assert hasattr(doc, 'metadata'), \
        f"Document instance should have 'metadata' attribute (type: {type(doc).__name__})"
    assert isinstance(doc.metadata, pdftract.Metadata), \
        f"Document.metadata should be Metadata instance, got {type(doc.metadata).__name__}"
    print("✓ Document.metadata is typed Metadata instance")

    assert doc.pages, "Document should contain pages"
    print("✓ Document.pages is not empty")

    # ===== Page-level type verification =====
    # Verify Page objects exist and are properly typed

    assert len(doc.pages) > 0, \
        f"Document should contain at least one page, found {len(doc.pages)} pages"

    # Verify each page is a Page instance
    for page in doc.pages:
        assert isinstance(page, pdftract.Page), \
            f"Each page should be a Page instance, got {type(page)}"
        assert hasattr(page, "spans"), "Page should have spans attribute"
        assert hasattr(page, "width"), "Page should have width attribute"
        assert hasattr(page, "height"), "Page should have height attribute"
    print("✓ Document has typed Page objects")

    # ===== Span-level type verification =====
    # Verify Span objects exist and are properly typed with expected attributes

    # Find a page with actual span content
    page_with_spans = None
    for page in doc.pages:
        if hasattr(page, 'spans') and len(page.spans) > 0:
            page_with_spans = page
            break

    assert page_with_spans is not None, \
        f"At least one page should have span content for type verification (checked {len(doc.pages)} page(s), fixture may be empty or malformed)"
    print(f"✓ Found page with {len(page_with_spans.spans)} span(s) for type checking")

    # Check each span is properly typed and has expected attributes
    for i, span in enumerate(page_with_spans.spans):
        assert isinstance(span, pdftract.Span), \
            f"spans[{i}] should be Span instance, got {type(span).__name__}"

        # Verify expected Span attributes exist (attribute presence checks)
        assert hasattr(span, 'text'), \
            f"spans[{i}] should have 'text' attribute (type: {type(span).__name__})"
        assert hasattr(span, 'bbox'), \
            f"spans[{i}] should have 'bbox' attribute (type: {type(span).__name__})"
        assert hasattr(span, 'font'), \
            f"spans[{i}] should have 'font' attribute (type: {type(span).__name__})"
        assert hasattr(span, 'size'), \
            f"spans[{i}] should have 'size' attribute (type: {type(span).__name__})"

        # Verify attribute types (type-specific validation)
        assert isinstance(span.text, str), \
            f"spans[{i}].text should be str, got {type(span.text).__name__} (expected: string content)"
        assert isinstance(span.bbox, tuple), \
            f"spans[{i}].bbox should be tuple, got {type(span.bbox).__name__} (expected: 4-element bounding box)"
        assert len(span.bbox) == 4, \
            f"spans[{i}].bbox should have 4 elements (x0,y0,x1,y1), got {len(span.bbox)} elements"
        assert isinstance(span.font, str), \
            f"spans[{i}].font should be str, got {type(span.font).__name__} (expected: font name)"
        assert isinstance(span.size, (int, float)), \
            f"spans[{i}].size should be numeric (int or float), got {type(span.size).__name__} (expected: font size)"

    print(f"✓ All {len(page_with_spans.spans)} span(s) properly typed with expected attributes (text, bbox, font, size)")

    # ===== Nested structure verification =====
    # Ensure complete type hierarchy integrity (Document -> Pages -> Spans)

    # Verify parent-child relationships: pages belong to the document
    total_pages = len(doc.pages)
    assert total_pages > 0, \
        f"Document should contain at least one page for hierarchy validation (Document->Pages->Spans chain), got {total_pages} pages"
    print(f"✓ Document owns {total_pages} page(s) (Document->Pages link valid)")

    # Verify each page is properly contained within the document structure
    for i, page in enumerate(doc.pages):
        assert isinstance(page, pdftract.Page), \
            f"doc.pages[{i}] should be a Page instance for hierarchy integrity, got {type(page).__name__}"
        # Verify page has structural attributes (page number, dimensions)
        assert hasattr(page, 'page'), \
            f"doc.pages[{i}] should have 'page' attribute (page number) for Page structure"
        assert hasattr(page, 'width'), \
            f"doc.pages[{i}] should have 'width' attribute for Page dimensions"
        assert hasattr(page, 'height'), \
            f"doc.pages[{i}] should have 'height' attribute for Page dimensions"

        # Verify Page.width is numeric (int or float)
        assert isinstance(page.width, (int, float)), \
            f"doc.pages[{i}].width should be numeric (int or float) for valid dimensions, got {type(page.width).__name__}"
    print(f"✓ All {total_pages} page(s) properly typed with structural attributes (page, width, height)")

    # Attribute access type verification
    # Verify Page.width is accessible and has correct numeric type
    assert hasattr(doc.pages[0], 'width'), \
        "Page instance should have 'width' attribute accessible"
    assert isinstance(doc.pages[0].width, (int, float)), \
        f"Page.width should be numeric (int or float), got {type(doc.pages[0].width).__name__}"
    print("✓ Page.width is accessible and has numeric type")

    # Verify Span.text is accessible and has correct string type
    assert hasattr(page_with_spans.spans[0], 'text'), \
        "Span instance should have 'text' attribute accessible"
    assert isinstance(page_with_spans.spans[0].text, str), \
        f"Span.text should be string, got {type(page_with_spans.spans[0].text).__name__}"
    print("✓ Span.text is accessible and has string type")

    # Verify at least one page has spans with real content
    pages_with_spans = [p for p in doc.pages if hasattr(p, 'spans') and len(p.spans) > 0]
    assert len(pages_with_spans) > 0, \
        f"At least one page should have spans populated, found {len(pages_with_spans)}/{total_pages} pages with spans (edge case: fixture may be empty)"
    print(f"✓ {len(pages_with_spans)}/{total_pages} page(s) have span content")

    # Verify span text contains real (non-empty) content, not placeholders
    total_spans = 0
    spans_with_text = 0
    for page in pages_with_spans:
        for span in page.spans:
            total_spans += 1
            if hasattr(span, 'text') and span.text:
                spans_with_text += 1

    # Edge case: no spans or all spans empty (indicates fixture problem)
    assert total_spans > 0, \
        f"Expected at least one span across {len(pages_with_spans)} page(s) with content, found {total_spans} spans (edge case: fixture may be malformed)"
    assert spans_with_text > 0, \
        f"Expected non-empty text content in spans, found {spans_with_text}/{total_spans} spans with text (edge case: spans may be placeholders)"
    print(f"✓ {spans_with_text}/{total_spans} span(s) have non-empty text content (real content)")

    # ===== COUNT INTEGRITY VERIFICATION =====
    # Verify count integrity: ensure all objects are properly accounted for
    total_pages_checked = len(doc.pages)
    total_spans_checked = sum(len(p.spans) for p in doc.pages if hasattr(p, 'spans'))

    # Verify page count consistency (detects structure corruption)
    assert total_pages_checked == total_pages, \
        f"Page count integrity check failed: expected {total_pages} pages, counted {total_pages_checked} pages (data structure may be corrupted)"

    # Verify span count consistency (detects lost/duplicate spans)
    assert total_spans_checked == total_spans, \
        f"Span count integrity check failed: expected {total_spans} spans, counted {total_spans_checked} spans (data structure may be corrupted)"

    print(f"✓ Count integrity verified: {total_pages} page(s), {total_spans} span(s), structure consistent")

    # ===== TEST SUCCESS SUMMARY =====
    print("\n" + "=" * 60)
    print("✅ ALL SMOKE TESTS PASSED")
    print("=" * 60)
    print(f"Document structure: {total_pages} page(s), {total_spans} span(s)")
    print(f"Content verification: {spans_with_text}/{total_spans} spans with text")
    print("Type contract verification: COMPLETE")
    print("All assertion types validated: isinstance, hasattr, length, type-specific")
    print("=" * 60)


def main() -> int:
    """Run the smoke test and return exit code."""
    print("=" * 60)
    print("pdftract SDK Smoke Test")
    print("=" * 60)
    print()

    try:
        test_extract_returns_typed_document()
        return 0
    except AssertionError as e:
        print(f"\n❌ Test failed: {e}")
        return 1
    except Exception as e:
        print(f"\n❌ Unexpected error: {e}")
        import traceback
        traceback.print_exc()
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: tests/smoke_test.rs
//! Smoke test for pdftract basic functionality
//!
//! This test verifies that the core PDF extraction pipeline works end-to-end.
//! It serves as a foundational smoke test that validates:
//! - Basic PDF parsing and loading
//! - Page extraction functionality
//! - Text extraction capabilities
//!
//! The test uses minimal fixtures to ensure fast execution and reliability.
//! This is the first test to run when validating the extraction pipeline.

use pdftract_core::{extract_pdf, ExtractionOptions, OutputOptions};
use std::path::Path;

#[test]
fn test_basic_pdf_extraction() {
    //! Verify basic PDF extraction works on a minimal fixture.
    //!
    //! This smoke test validates:
    //! - PDF file can be opened and parsed
    //! - At least one page is extracted
    //! - Extraction completes without errors
    //!
    //! Uses test-minimal.pdf (374 bytes) as a fast, reliable fixture.

    let fixture_path = Path::new("tests/fixtures/test-minimal.pdf");

    // Verify fixture exists
    assert!(
        fixture_path.exists(),
        "Test fixture should exist at {}",
        fixture_path.display()
    );

    // Run basic extraction
    let result = extract_pdf(
        fixture_path,
        &ExtractionOptions::default(),
        &OutputOptions::default(),
    );

    // Verify extraction succeeded
    assert!(
        result.is_ok(),
        "PDF extraction should succeed for minimal fixture: {:?}",
        result.err()
    );

    let extraction_result = result.unwrap();

    // Verify at least one page was extracted
    assert!(
        !extraction_result.pages.is_empty(),
        "At least one page should be extracted from the PDF"
    );

    // Verify we can access page data
    let first_page = &extraction_result.pages[0];
    assert!(
        first_page.width > 0.0,
        "First page should have a valid width"
    );
    assert!(
        first_page.height > 0.0,
        "First page should have a valid height"
    );
}

#[test]
fn test_sample_pdf_extraction() {
    //! Verify PDF extraction works on another simple fixture.
    //!
    //! Provides redundancy using sample.pdf (534 bytes) to ensure
    //! the extraction pipeline works across different minimal PDFs.

    let fixture_path = Path::new("tests/fixtures/sample.pdf");

    // Verify fixture exists
    assert!(
        fixture_path.exists(),
        "Test fixture should exist at {}",
        fixture_path.display()
    );

    // Run extraction
    let result = extract_pdf(
        fixture_path,
        &ExtractionOptions::default(),
        &OutputOptions::default(),
    );

    // Verify extraction succeeded
    assert!(
        result.is_ok(),
        "PDF extraction should succeed for sample fixture: {:?}",
        result.err()
    );

    let extraction_result = result.unwrap();

    // Verify at least one page was extracted
    assert!(
        !extraction_result.pages.is_empty(),
        "At least one page should be extracted from the PDF"
    );
}

#[test]
fn test_extract_returns_typed_document() {
    //! Verify complete nested structure integrity: Document -> Pages -> Spans.
    //!
    //! This comprehensive test validates the full type hierarchy and relationships
    //! throughout the PDF extraction pipeline. It ensures that the Document -> Pages -> Spans
    //! structure is properly maintained and all objects contain valid data.
    //!
    //! # Assertion Types Verified
    //!
    //! ## Fixture Validation
    //! - Fixture file exists at expected path
    //!
    //! ## Extraction Success
    //! - PDF extraction completes without errors
    //!
    //! ## Document-Level Checks
    //! - Document contains at least one page (page_count > 0)
    //! - All pages are accessible from the document
    //!
    //! ## Page-Level Checks
    //! - Each page has valid width > 0.0
    //! - Each page has valid height > 0.0
    //! - Spans are properly counted per page
    //!
    //! ## Span-Level Checks
    //! - Each span has bbox with exactly 4 coordinates
    //! - At least one span per page has non-empty text content
    //! - Text content is accessible via span.text
    //!
    //! ## Aggregate/Completeness Checks
    //! - At least one page has spans populated
    //! - Total span count > 0 across all pages
    //! - All objects in hierarchy are accounted for
    //!
    //! # Edge Cases Handled
    //! - Empty document (no pages): Error indicates expected vs actual page count
    //! - Pages without spans: Error identifies which page has empty spans
    //! - Spans without text: Error distinguishes between no spans and no text
    //! - Invalid page dimensions: Error shows page index and actual values
    //!
    //! # Test Fixture
    //! Uses `tests/fixtures/test-minimal.pdf` - a minimal PDF file optimized for
    //! fast, reliable testing. This fixture is known to contain at least one page
    //! with extractable text content.

    let fixture_path = Path::new("tests/fixtures/test-minimal.pdf");

    // ===== FIXTURE VALIDATION =====
    // Ensure the test fixture exists before attempting extraction
    assert!(
        fixture_path.exists(),
        "Test fixture should exist at {}, but file was not found. \
         Ensure tests/fixtures/test-minimal.pdf exists in the workspace.",
        fixture_path.display()
    );
    println!("✓ Fixture found: {}", fixture_path.display());

    // ===== EXTRACTION =====
    // Attempt to extract the PDF and capture the result
    let result = extract_pdf(
        fixture_path,
        &ExtractionOptions::default(),
        &OutputOptions::default(),
    );

    // Verify extraction succeeded (if this fails, the PDF may be corrupt)
    assert!(
        result.is_ok(),
        "PDF extraction should succeed for test-minimal.pdf. \
         Got error: {:?}. \
         This may indicate: corrupt PDF, unsupported PDF version, or parser bug.",
        result.err()
    );
    println!("✓ PDF extraction completed successfully");

    let extraction_result = result.unwrap();

    // ===== DOCUMENT-LEVEL CHECKS =====
    // Verify the document properly owns its pages and the hierarchy is intact
    let page_count = extraction_result.pages.len();
    assert!(
        page_count > 0,
        "Document should contain at least one page. \
         Expected: page_count > 0, \
         Actual: page_count = {}. \
         If no pages found, the PDF may be empty or the page parser failed.",
        page_count
    );
    println!("✓ Document contains {} page(s)", page_count);

    // Verify doc.pages is not empty
    assert!(
        !extraction_result.pages.is_empty(),
        "Document should contain pages"
    );

    // ===== PAGE-LEVEL CHECKS =====
    // These checks verify that each page has valid physical dimensions
    // and properly counts its associated spans.
    let mut total_spans = 0;
    let mut pages_with_spans = 0;

    for (page_idx, page) in extraction_result.pages.iter().enumerate() {
        // Verify page has valid dimensions
        assert!(
            page.width > 0.0,
            "Page {} should have valid width > 0. \
             Expected: width > 0.0, \
             Actual: width = {}. \
             Invalid dimensions indicate a page parsing error.",
            page_idx,
            page.width
        );
        assert!(
            page.height > 0.0,
            "Page {} should have valid height > 0. \
             Expected: height > 0.0, \
             Actual: height = {}. \
             Invalid dimensions indicate a page parsing error.",
            page_idx,
            page.height
        );

        // Count spans on this page
        let span_count = page.spans.len();
        total_spans += span_count;

        if span_count > 0 {
            pages_with_spans += 1;

            // ===== SPAN-LEVEL CHECKS =====
            // These checks verify that spans have valid structural data
            // (bounding boxes) and contain actual text content.
            let mut spans_with_text = 0;
            let mut spans_with_empty_text = 0;

            for (span_idx, span) in page.spans.iter().enumerate() {
                // Track spans with actual text content
                if !span.text.is_empty() {
                    spans_with_text += 1;
                } else {
                    spans_with_empty_text += 1;
                }

                // Verify span bounding box has exactly 4 coordinates [x0, y0, x1, y1]
                assert_eq!(
                    span.bbox.len(),
                    4,
                    "Page {} span {} should have bbox with exactly 4 coordinates [x0, y0, x1, y1]. \
                     Expected: bbox.len() == 4, \
                     Actual: bbox.len() = {}, value: {:?}. \
                     Invalid bbox dimensions indicate a span parsing error.",
                    page_idx,
                    span_idx,
                    span.bbox.len(),
                    span.bbox
                );

                // Verify bbox coordinates are finite (not NaN or infinite)
                assert!(
                    span.bbox[0].is_finite(),
                    "Page {} span {} should have finite x0 coordinate. \
                     Expected: finite value, \
                     Actual: x0 = {}. \
                     Non-finite coordinates indicate a span positioning error.",
                    page_idx,
                    span_idx,
                    span.bbox[0]
                );
                assert!(
                    span.bbox[1].is_finite(),
                    "Page {} span {} should have finite y0 coordinate. \
                     Expected: finite value, \
                     Actual: y0 = {}. \
                     Non-finite coordinates indicate a span positioning error.",
                    page_idx,
                    span_idx,
                    span.bbox[1]
                );
                assert!(
                    span.bbox[2].is_finite(),
                    "Page {} span {} should have finite x1 coordinate. \
                     Expected: finite value, \
                     Actual: x1 = {}. \
                     Non-finite coordinates indicate a span positioning error.",
                    page_idx,
                    span_idx,
                    span.bbox[2]
                );
                assert!(
                    span.bbox[3].is_finite(),
                    "Page {} span {} should have finite y1 coordinate. \
                     Expected: finite value, \
                     Actual: y1 = {}. \
                     Non-finite coordinates indicate a span positioning error.",
                    page_idx,
                    span_idx,
                    span.bbox[3]
                );
            }

            // At least one span on this page must have non-empty text
            assert!(
                spans_with_text > 0,
                "Page {} should have at least one span with non-empty text content. \
                 Expected: spans_with_text > 0, \
                 Actual: spans_with_text = {}, \
                 Total spans on page: {}, \
                 Spans with empty text: {}. \
                 If all spans are empty, the text extractor may have failed.",
                page_idx,
                spans_with_text,
                span_count,
                spans_with_empty_text
            );

            println!(
                "✓ Page {}: {} spans ({} with text, {} with empty text)",
                page_idx, span_count, spans_with_text, spans_with_empty_text
            );
        } else {
            println!("  Page {}: no spans found (edge case handled)", page_idx);
        }
    }

    // ===== AGGREGATE CHECKS =====
    // Verify overall document structure integrity

    // At least one page should have spans (non-empty content)
    assert!(
        pages_with_spans > 0,
        "At least one page should have spans populated. \
         Expected: pages_with_spans > 0, \
         Actual: pages_with_spans = {}, \
         Total pages: {}. \
         If no pages have spans, the PDF may contain no text or the span extractor failed.",
        pages_with_spans,
        page_count
    );
    println!("✓ {} page(s) have spans populated", pages_with_spans);

    // Total span count should be non-zero (completeness check)
    assert!(
        total_spans > 0,
        "Document should contain at least one span across all pages. \
         Expected: total_spans > 0, \
         Actual: total_spans = {}. \
         Zero spans indicates either an empty PDF or span extraction failure.",
        total_spans
    );
    println!("✓ Total spans in document: {}", total_spans);

    // Verify all pages are still accessible (no data loss during iteration)
    assert_eq!(
        extraction_result.pages.len(),
        page_count,
        "All pages should be accessible in the document (page count should remain stable). \
         Expected: {} pages, \
         Actual: {} pages. \
         Page count changed during iteration, indicating a possible data structure bug.",
        page_count,
        extraction_result.pages.len()
    );
    println!("✓ All {} pages accessible (no data loss)", page_count);
    println!("✓ All nested structure checks passed");
}
